### ISSUE TITLE
Figure hash table not updated with python setup.py test

### ISSUE THREAD

**Issue by ehsteve** [sunpy/sunpy#1472]
It only updates when running py.test explicitly.

**Comment by ayshih** [comment]
Your description is a bit ambiguous, but I'm pretty sure I know what happened on your machine.  Did the figure tests still pass even though you thought you deleted the hash table?

The issue here is that when you run `python setup.py test` instead of `py.test`, the SunPy code tree is copied over to a fixed directory under `build/` rather than a temporary directory.  This has the undesirable behavior that even when you remove a source file from your local repository, the file is still present in the `build/` directory from an earlier test run.  (@Cadair Is there a way to change this behavior?)  The workaround is to remove the hash table from the `build/` directory in addition to removing it from your local repository.

Note that this step of requiring the user to remove the hash table can be skipped if I change the implementation such that it always creates a new hash table for any mismatch.  However, I like having the user take the affirmative step to essentially ask for a new hash table, and it keeps pytest from creating a whole bunch of unnecessary hash tables in the temporary directory.

**Comment by bsipocz** [comment]
@ayshih - And is it normal that the failures show up even when starting from a clean build? (So I'm runnning `` python setup.py test``, but only after I've removed everything from previous builds?

**Comment by ayshih** [comment]
@bsipocz Can you elaborate on what you mean by "failures"?  Are the figure tests themselves failing, or are you trying and failing to create a new hash table?  You say you have failures when using `python setup.py test`, but how do things behave when you run `py.test`?

**Comment by bsipocz** [comment]
@ayshih - 4 tests fail. I'm checking with py.test now.

Here is the log (for ``python setup.py test``):
```
==================================================================== test session starts ====================================================================
platform darwin -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/sunpy-test-QzonhV, inifile: setup.cfg
plugins: capturelog-0.7, cov-2.1.0, html-1.6, mpl-0.3, selenium-1.0b1, variables-1.2
collected 840 items 

sunpy/database/tests/test_attrs.py .............................
sunpy/database/tests/test_caching.py ....
sunpy/database/tests/test_commands.py ...........................
sunpy/database/tests/test_database.py ................................................................
sunpy/database/tests/test_serialize.py ..........................
sunpy/database/tests/test_table.txt s
sunpy/database/tests/test_tables.py ...............
sunpy/image/tests/test_coalignment.py .............
sunpy/image/tests/test_resample.py .....
sunpy/image/tests/test_transform.py ...................................
sunpy/instr/tests/test_aia.py ..
sunpy/instr/tests/test_goes.py .....
sunpy/instr/tests/test_iris.py ..
sunpy/instr/tests/test_lyra.py ........
sunpy/io/tests/test_ana.py ......
sunpy/io/tests/test_filetools.py ..s.s....
sunpy/io/tests/test_fits.py ........
sunpy/io/tests/test_jp2.py sss
sunpy/lightcurve/tests/test_eve.py .
sunpy/lightcurve/tests/test_goes.py ..
sunpy/lightcurve/tests/test_lightcurve.py ........
sunpy/lightcurve/tests/test_noaa.py ..
sunpy/map/sources/tests/test_aia_source.py ....
sunpy/map/sources/tests/test_cor_source.py ....
sunpy/map/sources/tests/test_eit_source.py ....................................................
sunpy/map/sources/tests/test_euvi_source.py ....
sunpy/map/sources/tests/test_hmi_source.py ....
sunpy/map/sources/tests/test_iris_source.py ...
sunpy/map/sources/tests/test_lasco_source.py ....
sunpy/map/sources/tests/test_mdi_source.py ....
sunpy/map/sources/tests/test_sot_source.py .......
sunpy/map/sources/tests/test_swap_source.py ................
sunpy/map/sources/tests/test_xrt_source.py .....
sunpy/map/tests/test_header.py .
sunpy/map/tests/test_map_factory.py ..........
sunpy/map/tests/test_mapbase.py FFFF...................................................
sunpy/map/tests/test_mapcube.py ...
sunpy/net/jsoc/tests/test_attr.py ...........
sunpy/net/jsoc/tests/test_jsoc.py ..........
sunpy/net/tests/test_attr.py .
sunpy/net/tests/test_hek.py ............
sunpy/net/tests/test_helio.py .
sunpy/net/tests/test_util.py ...
sunpy/net/tests/test_vso.py ..................
sunpy/physics/transforms/tests/test_differential_rotation.py ...........
sunpy/physics/transforms/tests/test_solar_rotation.py ..
sunpy/roi/tests/test_chaincode.py ...........
sunpy/roi/tests/test_roi.py .........
sunpy/spectra/tests/test_callisto.py ...............
sunpy/spectra/tests/test_spectrogram.py ...................................
sunpy/sun/tests/test_sun.py ................
sunpy/tests/tests/test_helpers.py .
sunpy/tests/tests/test_main.py ......
sunpy/time/tests/test_julian.py ...............
sunpy/time/tests/test_time.py ............
sunpy/time/tests/test_timerange.py ..........................
sunpy/util/tests/test_cond_dispatch.py ......
sunpy/util/tests/test_datatype_factory_base.py ....
sunpy/util/tests/test_multimethod.py ..
sunpy/util/tests/test_scraper.py ..............
sunpy/util/tests/test_sysinfo.py .
sunpy/util/tests/test_util.py .............
sunpy/util/tests/test_xml.py ....................
sunpy/wcs/tests/test_wcs.py .........

========================================================================= FAILURES ==========================================================================
________________________________________________________________ test_plot_aia171_nowcsaxes _________________________________________________________________

aia171_test_map = SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]])

    
>   func = make_function_with_signature(func, name=name, **wrapped_args)

/Users/bsipocz/munka/devel/astropy/astropy/utils/decorators.py:679: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]]),)
kwargs = {}, name = 'sunpy.map.tests.test_mapbase.test_plot_aia171_nowcsaxes'
figure_hash = 'de14ec442bde91c5e486729b6639698cf1f6f2766056436ed943e33e4d929d80'

    @pytest.mark.figure
    @wraps(test_function)
    def wrapper(*args, **kwargs):
        plt.figure()
        name = "{0}.{1}".format(test_function.__module__, test_function.__name__)
        figure_hash = hash.hash_figure(test_function(*args, **kwargs))
        if name not in hash.hash_library:
            hash.hash_library[name] = figure_hash
            pytest.fail("Hash not present: {0}".format(name))
        else:
>           assert hash.hash_library[name] == figure_hash
E           assert '03845606cc7a...a711aa85e7602' == 'de14ec442bde9...3e33e4d929d80'
E             - 03845606cc7a61796bc333e8955067d40d40ee642b68ceda80da711aa85e7602
E             + de14ec442bde91c5e486729b6639698cf1f6f2766056436ed943e33e4d929d80

sunpy/tests/helpers.py:126: AssertionError
------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------
sunpy/map/mapbase.py:1440: Warning: This map is not properly oriented. Plot axes may be incorrect
  Warning)
---------------------------------------------------------------------- pytest-selenium ----------------------------------------------------------------------

_____________________________________________________________ test_plot_masked_aia171_nowcsaxes _____________________________________________________________

aia171_test_map = SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]])

    
>   func = make_function_with_signature(func, name=name, **wrapped_args)

/Users/bsipocz/munka/devel/astropy/astropy/utils/decorators.py:679: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]]),)
kwargs = {}, name = 'sunpy.map.tests.test_mapbase.test_plot_masked_aia171_nowcsaxes'
figure_hash = '5354a47baa8798801f8d34b3ee636f4ce2ef1ae1b2a94fb5ed9b0b2195dd0af0'

    @pytest.mark.figure
    @wraps(test_function)
    def wrapper(*args, **kwargs):
        plt.figure()
        name = "{0}.{1}".format(test_function.__module__, test_function.__name__)
        figure_hash = hash.hash_figure(test_function(*args, **kwargs))
        if name not in hash.hash_library:
            hash.hash_library[name] = figure_hash
            pytest.fail("Hash not present: {0}".format(name))
        else:
>           assert hash.hash_library[name] == figure_hash
E           assert '53dc4f9160a8...2efc98729e7d8' == '5354a47baa879...b0b2195dd0af0'
E             - 53dc4f9160a8862932c9c4dafd8aadb5c193db45886abc1b84c2efc98729e7d8
E             + 5354a47baa8798801f8d34b3ee636f4ce2ef1ae1b2a94fb5ed9b0b2195dd0af0

sunpy/tests/helpers.py:126: AssertionError
---------------------------------------------------------------------- pytest-selenium ----------------------------------------------------------------------

__________________________________________________________________ test_plot_masked_aia171 __________________________________________________________________

aia171_test_map = SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]])

    
>   func = make_function_with_signature(func, name=name, **wrapped_args)

/Users/bsipocz/munka/devel/astropy/astropy/utils/decorators.py:679: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]]),)
kwargs = {}, name = 'sunpy.map.tests.test_mapbase.test_plot_masked_aia171', figure_hash = 'cfe91a9486cfe421ddfe0d1d06eb7d28eac7815740461357d84835ac825cac35'

    @pytest.mark.figure
    @wraps(test_function)
    def wrapper(*args, **kwargs):
        plt.figure()
        name = "{0}.{1}".format(test_function.__module__, test_function.__name__)
        figure_hash = hash.hash_figure(test_function(*args, **kwargs))
        if name not in hash.hash_library:
            hash.hash_library[name] = figure_hash
            pytest.fail("Hash not present: {0}".format(name))
        else:
>           assert hash.hash_library[name] == figure_hash
E           assert '667cee5dab99...b0a5caf5732e7' == 'cfe91a9486cfe...835ac825cac35'
E             - 667cee5dab99cb53ff2ecf98f6eb8aa9addace0504268007696b0a5caf5732e7
E             + cfe91a9486cfe421ddfe0d1d06eb7d28eac7815740461357d84835ac825cac35

sunpy/tests/helpers.py:126: AssertionError
---------------------------------------------------------------------- pytest-selenium ----------------------------------------------------------------------

_____________________________________________________________________ test_plot_aia171 ______________________________________________________________________

aia171_test_map = SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]])

    
>   func = make_function_with_signature(func, name=name, **wrapped_args)

/Users/bsipocz/munka/devel/astropy/astropy/utils/decorators.py:679: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (SunPy AIAMap
---------
Observatory:	 SDO
Instrument:	 AIA 3
Detector:	 AIA
Mea...  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]]),)
kwargs = {}, name = 'sunpy.map.tests.test_mapbase.test_plot_aia171', figure_hash = '59d7f0c6b7febd2b9db81ff56d70e18d9f67a5d5488edb4424228851144da0cb'

    @pytest.mark.figure
    @wraps(test_function)
    def wrapper(*args, **kwargs):
        plt.figure()
        name = "{0}.{1}".format(test_function.__module__, test_function.__name__)
        figure_hash = hash.hash_figure(test_function(*args, **kwargs))
        if name not in hash.hash_library:
            hash.hash_library[name] = figure_hash
            pytest.fail("Hash not present: {0}".format(name))
        else:
>           assert hash.hash_library[name] == figure_hash
E           assert '651ea7e73553...660f0c735cc30' == '59d7f0c6b7feb...28851144da0cb'
E             - 651ea7e735537be5e4c50b1ceb46f4040411e074dd718951bba660f0c735cc30
E             + 59d7f0c6b7febd2b9db81ff56d70e18d9f67a5d5488edb4424228851144da0cb

sunpy/tests/helpers.py:126: AssertionError
---------------------------------------------------------------------- pytest-selenium ----------------------------------------------------------------------

============================================================ 105 tests deselected by '-k-online' ============================================================
==================================== 4 failed, 725 passed, 6 skipped, 105 deselected, 2 pytest-warnings in 60.58 seconds ====================================
```

**Comment by ayshih** [comment]
Hmm, those are outright figure-test failures (so are not actually related to this issue discussed here) and point to some difference in the plots you are generating.  You should see the exact same failures with `py.test`.  To answer your initial question, this is not "normal", and the fact that the Travis builds have been passing suggests that the difference is localized to your machine.  So, what does your `sunpy.util.system_info()` print out?

**Comment by ayshih** [comment]
Or...not?  The figure tests are failing on my machine too...  Needs more investigation.

**Comment by bsipocz** [comment]
@ayshih - py.test took much longer (about 15 minutes) and I have 6 failures with it, and only two of these image failures. It's very weird.

```
>>> sunpy.util.system_info() 
==========================================================
 SunPy Installation Information

==========================================================

###########
 General
###########
Time : Wednesday, 04. November 2015 10:05PM UT
System : Darwin
Processor : i386
Arch : 64bit
SunPy : 0.7.dev6338
SunPy_git : 176b9617922e9dc45357ed9ab8ad746c3fae3e68
OS: Mac OS X 10.10.5 (i386)


###########
 Required Libraries 
###########
Python: 2.7.10
NumPy: 1.10.1
SciPy: 0.16.0
matplotlib: 1.5.0
Astropy: 1.1.dev13982
Pandas: 0.17.0


###########
 Recommended Libraries 
###########
beautifulsoup: 4.4.0
PyQt: 4.11.3
SUDS: 0.4
Sqlalchemy: 1.0.8
Requests: 2.8.1
>>> 
```
Log for py.test is here: https://gist.github.com/bsipocz/16ad712d3829f554df4e

**Comment by Cadair** [comment]
I rebuilt the hashes for Travis recently. I wonder if something updated in the conda installs.

**Comment by bsipocz** [comment]
@Cadair - Travis is a different story, I can't see why it still works, all builds should have failed since this morning due to some conda changes (we had this problem in astropy and affiliates, and now I try to fix it everywhere, thus ended up here).

**Comment by ayshih** [comment]
@bsipocz it looks like running `py.test` like that skipped the two tests that are skipped if `wcsaxes` is not installed, which is why you only have two figure-test failures instead of four.

@Cadair I see where you changed the hashes (https://github.com/sunpy/sunpy/commit/84ae67ff3b14f8894aa86b871df70d606ddef2c4), but my setup (with an across-the-board-updated Anaconda) is still producing the old hashes...

**Comment by ayshih** [comment]
(That said, @bsipocz is getting that hashes that don't match the old ones *or* the new ones, so I don't know what's going on.)

**Comment by ayshih** [comment]
One theory: I'm testing on Windows, so maybe there's been some change to Anaconda such that the different platforms no longer produce the same hashes?  (That doesn't explain the failures that @bsipocz is seeing.)

**Comment by bsipocz** [comment]
@ayshih - Except that wcsaxes was installed. Both two runs were started from the same terminal after each other.

**Comment by bsipocz** [comment]
Also, I have the shiny new 1.5 matplotlib, can that cause problems?

**Comment by ayshih** [comment]
> @ayshih - Except that wcsaxes was installed. Both two runs were started from the same terminal after each other.

My suspicion is that our skip decorator may not work correctly when using `py.test` instead of `python setup.py test` because the former used to skip, and possibly still skips, some initial configuration code.

> Also, I have the shiny new 1.5 matplotlib, can that cause problems?

Oh, I'm sure it can!

**Comment by ayshih** [comment]
I've created a new issue (#1601) for the recent discussion since it is not directly related to the original issue.

**Comment by Cadair** [comment]
This is now working for me, if I delete the figure hashes they are being regenerated when I run `python setup.py test --figure` please reopen if you still have the issue.
